examples/ul2/src/utils/hf_prefixlm_converter.py

```diff
@@ -0,0 +1,769 @@
+                    combined_attention_mask, expanded_bidirectional_mask)
+
+        # [batch_size, seq_length] -> [batch_size, 1, tgt_length, src_length]
+        expanded_attn_mask = _expand_mask_bloom(attention_mask,
+                                                tgt_length=src_length)
+        combined_attention_mask = (expanded_attn_mask
+                                   if combined_attention_mask is None else
+                                   expanded_attn_mask | combined_attention_mask)
+
+        return combined_attention_mask
+
+    # Modified from transformers.models.bloom.modeling_bloom._prepare_alibi_transformer
+    def _build_alibi_tensor(
+        self: BloomModel,
+        batch_size: int,
+        query_length: int,
+        key_length: int,
+        dtype: torch.dtype,
+        device: torch.device,
+    ) -> torch.Tensor:
+        num_heads = self.config.n_head
+
+        closest_power_of_2 = 2**math.floor(math.log2(num_heads))
+        base = torch.tensor(2**(-(2**-(math.log2(closest_power_of_2) - 3))),
+                            device=device,
+                            dtype=torch.float32)
+        powers = torch.arange(1,
+                              1 + closest_power_of_2,
+                              device=device,
+                              dtype=torch.int32)
+        slopes = torch.pow(base, powers)
+
+        if closest_power_of_2 != num_heads:
+            extra_base = torch.tensor(
+                2**(-(2**-(math.log2(2 * closest_power_of_2) - 3))),
+                device=device,
+                dtype=torch.float32)
+            num_remaining_heads = min(closest_power_of_2,
+                                      num_heads - closest_power_of_2)
+            extra_powers = torch.arange(1,
+                                        1 + 2 * num_remaining_heads,
+                                        2,
+                                        device=device,
+                                        dtype=torch.int32)
+            slopes = torch.cat(
+                [slopes, torch.pow(extra_base, extra_powers)], dim=0)
+
+        qa = torch.arange(query_length, device=device,
+                          dtype=torch.int32).view(-1, 1)
+        ka = torch.arange(key_length, device=device,
+                          dtype=torch.int32).view(1, -1)
+        diffs = qa - ka + key_length - query_length
+        diffs = -diffs.abs()
+        alibi = slopes.view(1, num_heads, 1, 1) * diffs.view(
+            1, 1, query_length, key_length)
+        alibi = alibi.expand(batch_size, -1, -1,
+                             -1).reshape(-1, query_length, key_length)
+        return alibi.to(dtype)
+
+    # Modified from transformers.models.bloom.modeling_bloom.BloomModel.forward
+    def forward(
+        self: BloomModel,
+        input_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.Tensor, torch.Tensor],
+                                        ...]] = None,
```

Comment:
can we decompose this into a couple types with informative names? This is a looking little unreasonable + getting autoformatted poorly.

Thinking something like:
```python
KeyValueT = Tuple[torch.Tensor, torch.Tensor]
```
```python
past_key_values: Optional[Tuple[KeyValueT, ...]] = None,
```